src/mz390.java

```diff
@@ -7263,16 +7389,88 @@ private void exp_push_sdt(){
 				exp_stk_seta[tot_exp_stk_var-1] = seta_value; 
 				break;
 			case 'X': // X'C1' hex
-				seta_value = Long.valueOf(setc_value.substring(2,setc_value.length()-1),16).intValue();
-				exp_stk_seta[tot_exp_stk_var-1] = seta_value; 
+				if (isQuotedHexInteger(setc_value)){                                                           // #509
+					seta_value = Integer.parseUnsignedInt(setc_value.substring(2,setc_value.length()-1), 16);  // #509
+					exp_stk_seta[tot_exp_stk_var-1] = seta_value;                                              // #509
+				} else {                                                                                       // #509
+					log_error(195,"hex invalid self defining term - " + setc_value);                           // #509
+				}                                                                                              // #509
 				break;
 			case '*': // return max substring length
 				seta_value = max_substring_len; 
 				exp_stk_seta[tot_exp_stk_var-1] = seta_value;
 				break;	
 			default:  // must be ascii number
-				seta_value = get_int_from_string(setc_value,10); 
-				exp_stk_seta[tot_exp_stk_var-1] = seta_value;
+				if (is32BitBinaryInteger(setc_value, 10)){                           // #509
+					seta_value = get_int_from_string(setc_value,10);                 // #509
+					exp_stk_seta[tot_exp_stk_var-1] = seta_value;                    // #509
+				} else {                                                             // #509
+					/*
+					 * Special case: SDT started as "-2147483648". This
```

Comment:
Thanks for the extensive explanation. It really helps!

Reply:
You are welcome.